Mathlib/Analysis/InnerProductSpace/GramMatrix.lean

```diff
@@ -0,0 +1,125 @@
+-/
+
+open RCLike Real Matrix
+
+open scoped InnerProductSpace
+
+open scoped ComplexOrder
+
+variable {E n : Type*}
+variable {α : Type*}
+variable {𝕜 : Type*}
+namespace Matrix
+
+/-- The entries of a Gram matrix are inner products of vectors in an inner product space. -/
+def gram (𝕜 : Type*) [Inner 𝕜 E] (v : n → E) : Matrix n n 𝕜 := of fun i j ↦ inner 𝕜 (v i) (v j)
+
+local notation "⟪" x ", " y "⟫" => @inner 𝕜  _ _ x y
+
+/-- For the matrix `gram 𝕜 v`, the entry at `i j : n` equals `⟪v i, v j⟫`. -/
+lemma gram_apply [Inner 𝕜 E] {v : n → E} (i j : n) :
+    (gram _ v) i j = ⟪v i, v j⟫ := by
+  rw [gram, of_apply]
+
+/-- A Gram matrix is Hermitian. -/
+lemma isHermitian_gram [RCLike 𝕜]
+[SeminormedAddCommGroup E] [InnerProductSpace 𝕜 E] {v : n → E} : (gram 𝕜 v).IsHermitian := by
+  refine IsHermitian.ext_iff.mpr (fun i j ↦ ?_)
+  rw [gram, of_apply, of_apply]
+  simp only [RCLike.star_def, inner_conj_symm]
+
+theorem gram_eq [RCLike 𝕜]
+[SeminormedAddCommGroup E] [InnerProductSpace 𝕜 E]
+ [Fintype n] {v : n → E} (y : E) (x : n → 𝕜) (hy : y = ∑ i, x i • v i) :
+    ⟪y, y⟫ = star x ⬝ᵥ (gram 𝕜 v) *ᵥ x := by
+  calc
+  ⟪y, y⟫ = (∑ (i : n), ∑ (j : n), (starRingEnd 𝕜) (x i) * (x j) * ⟪v i, v j⟫) := by
+    simp_rw [hy, sum_inner, inner_sum, inner_smul_left, inner_smul_right, mul_assoc]
+  _ = (∑ (i : n), ∑ (j : n), (starRingEnd 𝕜) (x i) * (x j) * ((gram 𝕜 v) i j)) := by
+    simp_rw [gram_apply]
+  _ = star x ⬝ᵥ (gram 𝕜 v) *ᵥ x := by
+    simp_rw [dotProduct, mul_assoc, ← Finset.mul_sum, mulVec, dotProduct, mul_comm, ← star_def]
+    rfl
+
+theorem gram_eq' [RCLike 𝕜]
+[SeminormedAddCommGroup E] [InnerProductSpace 𝕜 E][Fintype n] {v : n → E} (x : n → 𝕜) :
+    ⟪∑ i, x i • v i, ∑ i, x i • v i⟫ = star x ⬝ᵥ (gram 𝕜 v) *ᵥ x := by
+  calc
+  ⟪∑ i, x i • v i, ∑ i, x i • v i⟫
+      = (∑ (i : n), ∑ (j : n), (starRingEnd 𝕜) (x i) * (x j) * ⟪v i, v j⟫) := by
+    simp_rw [sum_inner, inner_sum, inner_smul_left, inner_smul_right, mul_assoc]
+  _ = (∑ (i : n), ∑ (j : n), (starRingEnd 𝕜) (x i) * (x j) * ((gram 𝕜 v) i j)) := by
+    simp_rw [gram_apply]
+  _ = star x ⬝ᵥ (gram 𝕜 v) *ᵥ x := by
+    simp_rw [dotProduct, mul_assoc, ← Finset.mul_sum, mulVec, dotProduct, mul_comm, ← star_def]
+    rfl
+
+section PosSemidef
+
+variable [RCLike 𝕜] [SeminormedAddCommGroup E] [InnerProductSpace 𝕜 E]
+
+/-- A Gram matrix is positive semidefinite. -/
+theorem gram_posSemidef [Fintype n] {v : n → E} :
+    PosSemidef (gram 𝕜 v) := by
+  refine ⟨isHermitian_gram, fun x ↦ ?_⟩
+  rw [← gram_eq', le_iff_re_im]
+  simp only [map_zero, inner_self_im, and_true]
+  exact inner_self_nonneg
+
+example (P Q : Prop) : (P → Q) ↔ (¬Q → ¬ P) := by exact Iff.symm not_imp_not
+
+end PosSemidef
+
+section PosDef
+
+variable [RCLike 𝕜] [NormedAddCommGroup E] [InnerProductSpace 𝕜 E]
+
+/-- A Gram matrix `gram 𝕜 v` is positive definite iff `v` is linearly independent. -/
+theorem gram_posDef_iff_linearIndependent [Fintype n] {v : n → E} :
```

Comment:
Is the forward direction true under a weaker `SemiNormedAddCommGroup` assumption?